internal/serialization/object_data.go

```diff
@@ -121,10 +125,9 @@ func (o *ObjectDataOutput) WriteFloat64(v float64) {
 	o.position += bufutil.Float64SizeInBytes
 }
 
-func (o *ObjectDataOutput) WriteUTF(v string) {
+func (o *ObjectDataOutput) WriteString(v string) {
```

Comment:
Client protocol v3 string ser/deser is not compatible with v4. The changes are: first 4 byte now contain byte length, not the number of chars; the serialization now supports UTF-8 standard, incl. some 4-byte chars. The current implementation is also slow (at least, it shows in the `pprof` profile) and could be replaced with `[]bytes(v)`.

In general, the client should include binary compatibility tests in order to make sure that it plays nicely with other Hz clients.

See:
https://github.com/hazelcast/hazelcast/blob/master/hazelcast/src/test/java/com/hazelcast/nio/serialization/compatibility/BinaryCompatibilityTest.java
https://github.com/hazelcast/hazelcast-nodejs-client/blob/master/test/unit/serialization/BinaryCompatibilityTest.js

Reply:
Thanks for the pointers, I didn't yet deep dive in the serialization code. I'll fix that.

Reply:
Updated at https://github.com/yuce/hazelcast-go-client/commit/33bc62f5a109f751db1af984138d5fa320da3730 and https://github.com/yuce/hazelcast-go-client/commit/17b753d61dc46622c9baad39e079a014db781f8f

Reply:
Resolving this one since the string serialization is now fixed. Still, it makes sense to plan adding binary compatibility tests for the client.